### PR TITLE
Srizvi/react router/#143336083

### PR DESCRIPTION
**Pull request by srizvi-official**
Basic react routing mechanism in place, it also include some code refactoring. But I want you guys to checkout this branch and review this. When we all agreed to that implementation, we will move forward in that direction, as its crucial and require agreement.

There are some items which I need to figure out. But, pushing this branch to github from my local so that you guys can contribute on it.

Although we have working routing in place, we can set up client side routing as well as match server side routing and render specific components accordingly.

But I am looking for better compatibility with store (redux), how we can add and manipulate various variables set in `windows.Onbionic.<some-vars>` in that name space than dispatch that to store.

Any idea's around that highly appreciated.

### CODE REVIEW

**Comment by mckinley** [comment]
the `window.BIONIC...` variable is only there to avoid an initial request to the server for our first data load. All later changes to redux will be through ajax request, but they will operate very similarly.

After the initial data load from the page variable, all data loads function as normal, through ajax requests.

**Comment by srizvi-official** [comment]
Look current implementation it works that way

**Comment by srizvi-official** [comment]
I am going to use [React Refetch](https://github.com/heroku/react-refetch) that will incorporate a data fetch from the server on particular component, routed (client-side) respectively from React Router. Currently it will mimic the data fetch from `window.BIONIC` object as per current implementation. I am also working to resolve the client-side route that doesn't exist on rails server.

**Comment by srizvi-official** [comment]
Please don't merge this today, we will do that in morning meanwhile there are some thing I wanted to test on my end,

Routes working fine but if we have any routes just on client side than it working fine unless you refresh that browser in that case rails routing invoke and error out that route as its not configured in rails.

Thats because our client app working under port 3000 listen by rails so each page refresh invoke rails and routes get failed.

Without refreshing on client side its working fine, even working fine in case when we have rails route in place on back-end as well as front-end. Its just issue with pure client side routing. Give me a chance to look into this today

**Comment by mckinley** [comment]
Seems like we should just have a catch-all rails route that shows the react app. And then 404s are handled by react. Same as when we separate client and api.

**Comment by srizvi-official** [comment]
Thinking along the same lines, let me in than we discuss and move forward in that direction

File: package.json [hunk 1]
```diff
@@ -29,7 +29,7 @@
     "react": "^15.4.2",
     "react-dom": "^15.4.2",
     "react-redux": "^5.0.3",
-    "react-router": "^4.0.0",
+    "react-router": "^2.7.0",
```
Comment (mckinley) [inline package.json]: Did you run into versioning problems here? Just curious about the downgrade. wanted to make sure it is intentional.

File: app/javascript/react/components/OpportunityAreaHeader/OpportunityArea.js [hunk 2]
```diff
@@ -1,40 +1,24 @@
...
-            <h5 className="oa-cont" title={name}>{name}</h5>
-            <div className="bud-cont"><span className="box-label">budget</span>
-              <span className="box-content">{budgetFormatted}</span></div>
-            <div className="bar-container">
-              <div className="total-budget" title={budgetTotalFormatted} style={budgetTotalStyle}/>
-              <div className="spend-budget" title={spendTotalFormatted} style={spendTotalStyle}/>
-            </div>
-          </div>
-        </a>
+        <div style={{backgroundColor: color}}>color: {color} </div>
```
Comment (mckinley) [inline app/javascript/react/components/OpportunityAreaHeader/OpportunityArea.js]: Looks like this is a regression. We lost the appropriate markup.

File: package.json (hunk 1 above)
Comment (srizvi-official) [inline package.json]: Yes I add that dependency as part of that PR, first I use latest react-router but, it has some compatibility issue so I downgraded as new one released just 19 hours ago. So I tend to use stable version.

image

File: app/javascript/react/components/OpportunityAreaHeader/OpportunityArea.js (hunk 2 above)
Comment (srizvi-official) [inline app/javascript/react/components/OpportunityAreaHeader/OpportunityArea.js]: Yes. Will be fixed in the next commit. I lost in in refactoring directory structure and renaming OpportunityAreaHeader/index

**Comment by srizvi-official** [comment]
@bronson please, read it https://github.com/heroku/react-refetch

Thats good for read heavy application

[2:03] 
build by heroku

[2:04] 
This project was inspired by (and forked from) React Redux. Redux/Flux is a wonderful library/pattern for applications that need to maintain complicated client-side state; however, if your application is mostly fetching and rendering read-only data from a server, it can over-complicate the architecture to fetch data in actions, reduce it into the store, only to select it back out again. The other approach of fetching data inside the component and dumping it in local state is also messy and makes components smarter and more mutable than they need to be. This module allows you to wrap a component in a connect() decorator like react-redux, but instead of mapping state to props, this lets you map props to URLs to props. This lets you keep your components completely stateless, describe data sources in a declarative manner, and delegate the complexities of data fetching to this module. Advanced options are also supported to lazy load data, poll for new data, and post data to the server . (edited)

[2:06] 
btw - I am going to create PR for react routing its good to go but if we later wanted to incorporate above library I can refactor application code accordingly (if we mutually find useful for our app)

**Comment by srizvi-official** [comment]
HeY @bronson there

react routing stuff done, responses to your comments over PR
only blocker is which I am unable to figure out. One spec failing error like some class not found
I observed when I merged changes from master to my branch start getting this
other than that I tested its look good
above you can see long comment of mine, need your opinion on that
I wanted to merge that PR to master if all looks good
only challenge that spec failing
That will be really helpful if you fixed that spec thing because m already late have to leave for home

**Comment by srizvi-official** [comment]
image


Randomized with seed 35654
*...........................F

Pending: (Failures listed here are expected and do not affect your suite's status)

  1) Partner add some examples to (or delete) /var/www/html/bionic-dashboard/spec/models/partner_spec.rb
     # Not yet implemented
     # ./spec/models/partner_spec.rb:4


Failures:

  1) PortfolioHeader data is displayed
     Failure/Error:
       within '.portfolio-header' do
         expect(page).to have_content portfolio.name.upcase
         expect(page).to have_content projects.count
         expect(page).to have_content opportunity_areas.count
         expect(page).to have_content portfolio.portfolio_updates.count
         expect(page).to have_content "$#{(portfolio.revenue_target.to_f / 1_000_000).round}M"
         expect(page).to have_content "$#{(portfolio.budget_recapture.to_f / 1_000_000).round}M"
         expect(page).to have_content 'INVESTMENT SUMMARY'
         expect(page).to have_content 'VALIDATE'
         expect(page).to have_content 'TEAM'
     
     Capybara::ElementNotFound:
       Unable to find css ".portfolio-header"
     # ./spec/features/portfolio_header_spec.rb:19:in `block (3 levels) in <top (required)>'

Finished in 9.73 seconds (files took 4.02 seconds to load)
29 examples, 1 failure, 1 pending

Failed examples:

rspec ./spec/features/portfolio_header_spec.rb:17 # PortfolioHeader data is displayed

**Comment by srizvi-official** [comment]
that's what I am getting, thats appearing after merge master into my branch so source not identified yet. If you do that really helpful otherwise I fixed that first thing on my next working day.

**Comment by mckinley** [comment]
Its all fixed up and merged. We can talk about it at standup.

**Comment by srizvi-official** [comment]
Thanks @mckinley

**Pull request merged**